Implementation for android (maybe using termux)

Is it possible?

For the compile step, these diff makes it almost work 
bandwhich_diff
```diff
diff --git a/src/os/mod.rs b/src/os/mod.rs
index 056d44e..374fdbc 100644
--- a/src/os/mod.rs
+++ b/src/os/mod.rs
@@ -1,4 +1,4 @@
-#[cfg(target_os = "linux")]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 pub(self) mod linux;
 
 #[cfg(any(target_os = "macos", target_os = "freebsd"))]
diff --git a/src/os/shared.rs b/src/os/shared.rs
index 462f269..bcef6ea 100644
--- a/src/os/shared.rs
+++ b/src/os/shared.rs
@@ -11,7 +11,7 @@ use ::std::time;
 use crate::os::errors::GetInterfaceErrorKind;
 use signal_hook::iterator::Signals;
 
-#[cfg(target_os = "linux")]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 use crate::os::linux::get_open_sockets;
 #[cfg(any(target_os = "macos", target_os = "freebsd"))]
 use crate::os::lsof::get_open_sockets;
@@ -245,7 +245,7 @@ fn eperm_message() -> &'static str {
 }
 
 #[inline]
-#[cfg(target_os = "linux")]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 fn eperm_message() -> &'static str {
     r#"
     Insufficient permissions to listen on network interface(s). You can work around
```

procfs_diff
```diff
diff --git a/src/process/mod.rs b/src/process/mod.rs
index 0c30c22..6915316 100644
--- a/src/process/mod.rs
+++ b/src/process/mod.rs
@@ -54,12 +54,12 @@
 
 use super::*;
 use crate::from_iter;
-
+use std::convert::TryFrom;
 use std::ffi::OsString;
 use std::fs;
 use std::io::{self, Read};
-#[cfg(unix)]
-use std::os::linux::fs::MetadataExt;
+#[cfg(linux)]
+use std::os::unix::fs::MetadataExt;
 use std::path::PathBuf;
 use std::str::FromStr;
 
@@ -90,6 +90,9 @@ impl FakeMedatadataExt for std::fs::Metadata {
     }
 }
 
+#[cfg(target_os = "android")]
+use std::os::android::fs::MetadataExt;
+
 bitflags! {
     /// Kernel flags for a process
     ///
@@ -179,9 +182,9 @@ bitflags! {
 bitflags! {
     /// The mode (read/write permissions) for an open file descriptor
     pub struct FDPermissions: u32 {
-        const READ = libc::S_IRUSR;
-        const WRITE = libc::S_IWUSR;
-        const EXECUTE = libc::S_IXUSR;
+        const READ = libc::S_IRUSR as u32;
+        const WRITE = libc::S_IWUSR as u32;
+        const EXECUTE = libc::S_IXUSR as u32;
     }
 }
 
@@ -760,7 +763,7 @@ impl Process {
                 let link_os: &OsStr = link.as_ref();
                 vec.push(FDInfo {
                     fd,
-                    mode: md.st_mode() & libc::S_IRWXU,
+                    mode: md.st_mode() & u32::try_from(libc::S_IRWXU).unwrap(),
                     target: expect!(FDTarget::from_str(expect!(link_os.to_str()))),
                 });
             }
```

with these changes it fails in linking:
```
  = note: pnet_datalink-0.26.0/src/unix_interfaces.rs:39: error: undefined reference to 'getifaddrs'
             pnet_datalink-0.26.0/src/unix_interfaces.rs:74: error: undefined reference to 'freeifaddrs'
```

hacking those lines out makes it compile and have a runtime error as expected, so this should be the last issue regarding compiling.
I'm hoping someone with more knowledge about the subject can figure whats the next step is, as far as I can see those functions exists in `arm-linux-androideabi` https://rust-lang.github.io/libc/arm-linux-androideabi/libc/fn.getifaddrs.html

Note: compile command: `cross build --target arm-linux-androideabi` [cargo-cross](https://github.com/rust-embedded/cross)

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -47,12 +47,24 @@ jobs:
           toolchain: ${{ matrix.rust }}
           components: rustfmt, clippy
 
+      - name: Install Cross on Ubuntu
+        if: matrix.os == 'ubuntu-latest'
+        # The latest realese of `cross` is not able to build/link for `aarch64-linux-android`
+        # See: https://github.com/cross-rs/cross/issues/1222
+        # This is fixed on `main` but not yet released. To avoid a breakage somewhen in the future
+        # pin the cross revision used to the latest HEAD at 04/2024. 
+        # Remove the git source and revision once cross 0.3 is released.
+        run: cargo install --git https://github.com/cross-rs/cross.git --rev 085092c cross
+
       - name: Check formatting
         run: cargo fmt --all -- --check
 
       - name: Build
-        id: build
         run: cargo build --verbose
+
+      - name: Build target aarch64-linux-android
+        if: matrix.os == 'ubuntu-latest'
+        run: cross build --target aarch64-linux-android --verbose
 
       # This is useful for debugging problems when the expected build artifacts
       # (like shell completions and man pages) aren't generated.
@@ -102,6 +114,13 @@ jobs:
         with:
           name: ${{ matrix.os }}-${{ matrix.rust }}-failed_snapshots
           path: '**/*.snap.new'
+
+      - name: Upload android binary
+        if: ${{ matrix.os == 'ubuntu-latest' && ( success() || steps.build.outcome == 'success' ) }}
+        uses: actions/upload-artifact@v3
+        with:
+          name: aarch64-linux-android-${{ matrix.rust }}
+          path: target/aarch64-linux-android/debug/bandwhich
 
       - name: Upload unix binary
         if: ${{ matrix.os != 'windows-latest' && ( success() || steps.build.outcome == 'success' ) }}

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -52,11 +52,17 @@ jobs:
     strategy:
       matrix:
         build:
+          - aarch64-linux-android
           - linux-x64-gnu
           - linux-x64-musl
           - macos-x64
           - windows-x64-msvc
         include:
+          - cargo: cargo # default; overwrite with `cross` if necessary
+          - build: aarch64-linux-android
+            os: ubuntu-latest
+            target: aarch64-linux-android
+            cargo: cross
           - build: linux-x64-gnu
             os: ubuntu-latest
             target: x86_64-unknown-linux-gnu
@@ -84,13 +90,17 @@ jobs:
         if: matrix.target == 'x86_64-unknown-linux-musl'
         run: sudo apt-get install -y --no-install-recommends musl-tools
 
+      - name: Install cross
+        if: matrix.cargo == 'cross'
+        run: cargo install --git https://github.com/cross-rs/cross.git --rev 085092c cross
+
       - name: Build release binary
         shell: bash
         env:
           RUSTFLAGS: "-C strip=symbols"
         run: |
           mkdir -p "$BANDWHICH_GEN_DIR"
-          cargo build --verbose --release --target ${{ matrix.target }}
+          ${{ matrix.cargo }} build --verbose --release --target ${{ matrix.target }}
 
       - name: Collect build artifacts
         shell: bash

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 ## [Unreleased]
 
 ## Fixed
+
 * Remove redundant imports #377 - @cyqsimon
 * CI: use GitHub API to exempt dependabot from changelog requirement #378 - @cyqsimon
 * Remove unnecessary logging synchronisation #381 - @cyqsimon
@@ -15,10 +16,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 * Support build for `target_os`` `android` #384 - @flxo
 
 ## Added
+
 * CI: include generated assets in release archive #359 - @cyqsimon
 * Add PID column to the process table #379 - @notjedi
+* CI: add builds for target `aarch64-linux-android` #384 - @flxo
 
 ## Changed
+
 * CI: strip release binaries for all targets #358 - @cyqsimon
 * Bump MSRV to 1.74 (required by clap 4.5; see #373)
 * CI: Configure dependabot grouping #395 - @cyqsimon

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 * Remove unnecessary logging synchronisation #381 - @cyqsimon
 * Apply suggestions from new clippy lint clippy::assigning_clones #382 - @cyqsimon
 * Fix IPv6 socket detect logic #383 - @cyqsimon
+* Support build for `target_os`` `android` #384 - @flxo
 
 ## Added
 * CI: include generated assets in release archive #359 - @cyqsimon

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -43,7 +43,7 @@ trust-dns-resolver = "0.23.2"
 unicode-width = "0.1.11"
 strum = { version = "0.26.2", features = ["derive"] }
 
-[target.'cfg(target_os = "linux")'.dependencies]
+[target.'cfg(any(target_os = "android", target_os = "linux"))'.dependencies]
 procfs = "0.16.0"
 
 [target.'cfg(any(target_os = "macos", target_os = "freebsd"))'.dependencies]

--- a/README.md
+++ b/README.md
@@ -56,7 +56,7 @@ For more details, see [The Future of Bandwhich #275](https://github.com/imsnif/b
 
 ### Download a prebuilt binary
 
-If you're on linux, you can download the generic binary from the releases.
+If you're on `android` or `linux`, you can download the generic binary from the [releases](https://github.com/imsnif/bandwhich/releases)
 
 ### Arch Linux
 
@@ -122,6 +122,19 @@ The minimum supported Rust version is **1.70.0**.
 ```
 cargo install bandwhich
 ```
+
+#### Building from source for Android (`aarch64-linux-android`)
+
+Building for target `aarch64-linux-android` is supported via [cross](https://github.com/cross-rs/cross). Until [#1222](https://github.com/cross-rs/cross/issues/1222) is solved, use the latest HEAD:
+
+```sh
+cargo install --git https://github.com/cross-rs/cross.git cross
+cross build --release --target aarch64-linux-android
+```
+
+Kindly be aware that this process generates a pure binary file and not an APK suitable for installation on any arbitrary Android device.
+
+The Android support for bandwhich beyond its build is *not* endorsed by this project and may be unstable or non-functional.
 
 ### OpenWRT
 

--- a/src/os/mod.rs
+++ b/src/os/mod.rs
@@ -1,4 +1,4 @@
-#[cfg(target_os = "linux")]
+#[cfg(any(target_os = "android", target_os = "linux"))]
 mod linux;
 
 #[cfg(any(target_os = "macos", target_os = "freebsd"))]

--- a/src/os/shared.rs
+++ b/src/os/shared.rs
@@ -13,7 +13,7 @@ use tokio::runtime::Runtime;
 
 use crate::{network::dns, os::errors::GetInterfaceError, OsInputOutput};
 
-#[cfg(target_os = "linux")]
+#[cfg(any(target_os = "android", target_os = "linux"))]
 use crate::os::linux::get_open_sockets;
 #[cfg(any(target_os = "macos", target_os = "freebsd"))]
 use crate::os::lsof::get_open_sockets;
@@ -220,7 +220,7 @@ fn eperm_message() -> &'static str {
 }
 
 #[inline]
-#[cfg(target_os = "linux")]
+#[cfg(any(target_os = "android", target_os = "linux"))]
 fn eperm_message() -> &'static str {
     r#"
     Insufficient permissions to listen on network interface(s). You can work around

